### PR TITLE
Track the parent block to optimize hierarchy selectors

### PR DESCRIPTION
**Pull request by youknowriad**
Extracted from #16368

This PR tracks the parent of a block in the "blocks.parents" state tree in the reducer to optimizer the performance of some block editor selectors.

On its own, this PR doesn't seem to have a huge impact on performance bit it's a good step towards #16368

### CODE REVIEW

File: packages/block-editor/src/store/reducer.js
Comment (aduth) [inline packages/block-editor/src/store/reducer.js]: Seems like a good use-case for `Array#reduce`.

```js
return blocks.reduce( ( result, block ) => Object.assign(
	result,
	{ [ block.clientId ]: rootClientId },
	mapBlockParents( block.innerBlocks, block.clientId )
), {} );
```

File: packages/block-editor/src/store/reducer.js
Comment (aduth) [inline packages/block-editor/src/store/reducer.js]: I feel we could be more specific with this comment, particularly around what "performance reasons" there are.

```suggestion
	// While technically redundant data as the inverse of `order`, it serves as
	// an optimization for the selectors which derive the ancestry of a block.
```

File: packages/block-editor/src/store/reducer.js
Comment (aduth) [inline packages/block-editor/src/store/reducer.js]: Similarly, we could be more specific to guide future maintainers on what the memory leak is:

```suggestion
					// This should include the parents to avoid a memory leak
					// caused by orphaned references to blocks.
```

Comment (aduth) [reply packages/block-editor/src/store/reducer.js]: This behavior seems particularly important to unit test.

Comment (youknowriad) [reply packages/block-editor/src/store/reducer.js]: I think I want to solve this before merge

**Comment by youknowriad** [comment]
I fixed the memory leak in the parents reducer. It turns out that I discovered memory leaks in other reducers (byClientId and order), I won't be fixing those right now in this PR though. A generic fix would be good but it's not possible due to the fact that we track reusable blocks their as well.

**Comment by youknowriad** [comment]
I think this should be ready and I'm noticing something between 2-5% performance improvement.

File: packages/block-editor/src/store/reducer.js
Comment (aduth) [inline packages/block-editor/src/store/reducer.js]: Is there a specific reason we need to rename these properties from `clientIds`, rather than just "enhancing" the existing value to account for the cascade? Seems like it might make this a bit simpler to implement, since you wouldn't need the `switch` to vary the behavior.

File: packages/block-editor/src/store/reducer.js
Comment (aduth) [inline packages/block-editor/src/store/reducer.js]: Re: My previous comment, I guess we still reference the original `clientIds` here. Can you clarify what you're doing here in referencing only the first of the original of the `clientIds` ?

File: packages/block-editor/src/store/test/reducer.js
Comment (aduth) [inline packages/block-editor/src/store/test/reducer.js]: 😆

File: packages/block-editor/src/store/selectors.js
Comment (aduth) [inline packages/block-editor/src/store/selectors.js]: I don't know if you have feelings one way or the other, but the stronger convention in this file is to use a function declaration if it's not a memoized selector (mentioning partly because it's easy to overlook when unwrapping as we're doing here).

File: packages/block-editor/src/store/reducer.js
Comment (aduth) [inline packages/block-editor/src/store/reducer.js]: Is this condition related to your comment at https://github.com/WordPress/gutenberg/pull/16392#issuecomment-507735698 with regards to orphaned references? Should we comment it as a temporary fix if it is meant to be temporary?

File: packages/block-editor/src/store/reducer.js
Comment (aduth) [inline packages/block-editor/src/store/reducer.js]: While we're simply inheriting precedent here, I wonder if we could optimize this to only create a shallow clone of the array if the loop below would result in an `Array#push`. Seems more likely most blocks would not need to modify the original set (optimizing for non-nested blocks).

We do similar in many of our reducers for objects:

```js
let result = clientIds;

for ( let i = 0; i < result.length; i++ ) {
	if ( ! state.order[ result[ i ] ] ) {
		continue;
	}

	if ( result === clientIds ) {
		result = [ ...result ];
	}

	result.push( ...state.order[ result[ i ] ] );
}

return result;
```

I kinda doubt it would have much impact, and questionable whether the added "complexity" is worthwhile. That said, considering we already have prior art with `getMutateSafeObject` in this function, we could expand it to cover this use-case.

(Not a blocker)

File: packages/block-editor/src/store/reducer.js
Comment (youknowriad) [reply packages/block-editor/src/store/reducer.js]: I've been caught off guard in several occasions because of this. The fact that the property don't contain the same value between reducers (selection and blocks), and also the difference between the action creator (which I tend to take a look at to understand the action) led me to this change. 

It is still a different action but instead but I think the fact that we introduce a new key makes it less confusing.

File: packages/block-editor/src/store/reducer.js
Comment (youknowriad) [reply packages/block-editor/src/store/reducer.js]: Basically we need the parent of the replaced blocks to assign it as parent of the inserted blocks. `replacedClientIds` contains more than just the "root" level which in theory could result in the wrong value here.

File: packages/block-editor/src/store/selectors.js
Comment (youknowriad) [reply packages/block-editor/src/store/selectors.js]: I don't mind changing. My brain doesn't use a consistent method yet :P

File: packages/block-editor/src/store/reducer.js
Comment (youknowriad) [reply packages/block-editor/src/store/reducer.js]: no, to be completely honest, I tried but failed to find why this condition is necessary. When you remove an "empty paragraph" using backspace, there's a JS error that is triggered if you remove this condition.

File: packages/block-editor/src/store/reducer.js
Comment (mcsf) [reply packages/block-editor/src/store/reducer.js]: I get the motivation for this. It could, however, be pretty jarring for someone who notices the discrepancy and pulls up the definition of `removeBlocks`, since it clearly only returns `clientIds` in its payload.

Perhaps update `removeBlocks` to make that relationship clearer?

```js
// actions.js
yield {
	type: 'REMOVE_BLOCKS',
	clientIds,
	removedClientIds: undefined // Explain bla bla higher-order reducer
};
```

In alternative, if we're really strict about action types, it would be more "correct" to let this higher-order reducer divert from one action type to a new one:

```js
// reducer.js
if ( state ) {
	switch ( action.type ) {
		case 'REMOVE_BLOCKS':
			action = {
				type: DERIVED_TYPE_SIMILAR_TO_REMOVE_BLOCKS,
				removedClientIds: …
```

Comment (aduth) [reply packages/block-editor/src/store/reducer.js]: Alternatively, with controls implementation, these days we could probably incorporate this into the action creator itself, rather than as a reducer enhancer. That way the logic is consolidated to one place.

**Review by aduth — APPROVED** [review]
I'm not sure if you wanted to make revisions based on @mcsf 's comment (I think either could maybe help), but in any case this looks like a solid improvement as it is.

**Pull request merged**
